packages/@react-aria/grid/src/useGridCell.ts

```diff
@@ -205,18 +249,101 @@ export function useGridCell<T, C extends GridCollection<T>>(props: GridCellProps
+        if (keyboardNavigationBehavior === 'arrow' && !e.altKey && ref.current.contains(e.target as Element)) {
           e.stopPropagation();
           e.preventDefault();
-          ref.current.parentElement.dispatchEvent(
-            new KeyboardEvent(e.nativeEvent.type, e.nativeEvent)
-          );
+
+          if (isEditing) {
+            ref.current.dispatchEvent(
+              new KeyboardEvent(e.nativeEvent.type, {...e.nativeEvent, key: e.key === 'ArrowUp' ? 'ArrowLeft' : 'ArrowRight'})
+            );
+          } else {
+            ref.current.parentElement.dispatchEvent(
+              new KeyboardEvent(e.nativeEvent.type, e.nativeEvent)
+            );
+          }
+        }
+        break;
+      }
+      case 'F2':
+      case 'Enter': {
+        if (isEditable) {
+          let focusable = walker.nextNode() as FocusableElement;
+
+          if (focusable && document.activeElement === ref.current) {
+            state.setEditingKey(node.key);
+            focusSafely(focusable);
+          }
+        }
+
+        if (isEditable && state.editingKey && (isSingleLineTextInput(document.activeElement) || e.key === 'F2')) {
+          state.setEditingKey(null);
+          focusSafely(ref.current);
+        }
+        break;
+      }
+      case 'Escape': {
+        if (state.editingKey || keyboardNavigationBehavior === 'tab') {
+          state.setEditingKey(null);
+          focusSafely(ref.current);
+        }
+        break;
+      }
+      case 'Tab': {
+        if (keyboardNavigationBehavior === 'tab') {
+          // If there is another focusable element within this item, stop propagation so the tab key
+          // is handled by the browser and not by useSelectableCollection (which would take us out of the list).
+          let walker = getFocusableTreeWalker(ref.current, {tabbable: true});
+          walker.currentNode = document.activeElement;
+          let next = e.shiftKey ? walker.previousNode() : walker.nextNode();
+
+          if (next || isEditing) {
+            e.stopPropagation();
+
+            if (isEditable && document.activeElement === ref.current) {
+              state.setEditingKey(node.key); // Do we want this?
+            }
+
+            let [start, end] = wrap(walker);
+
+            if (isEditing && !e.shiftKey && document.activeElement === end) {
+              e.preventDefault();
+              focusSafely(start);
+            }
+
+            if (isEditing && e.shiftKey && document.activeElement === start) {
+              e.preventDefault();
+              focusSafely(end);
+            }
+          }
+        }
+        break;
+      }
+      default: {
+        if (document.activeElement === ref.current && /^[a-zA-Z0-9]$/.test(e.key)) {
+          let focusable = first(walker);
+
+          if (focusable) {
+            state.setEditingKey(node.key);
+            focusSafely(focusable);
+          }
         }
         break;
+      }
+    }
+  };
+
+  let onBlur = (e) => {
```

Comment:
Known issue: Edit mode is not recognized when clicking into the editable cell.

It likely makes sense to enter edit mode `onFocus` instead of `onKeyDown`, similar to how we exit `onBlur`.

Reply:
If I remember correctly, we did make use of `onFocus` for exactly this in our old edit mode explorations. That combined with some visual affordances via the `data-editing` that you added should hopefully be enough indication to the user that they are in edit mode and thus will need to exit it via Escape